src/crimson/os/seastore/transaction_manager.h

```diff
@@ -362,6 +362,73 @@ class TransactionManager : public AsyncCleaner::ExtentCallbackInterface {
     });
   }
 
+  /**
+   * map_existing_extent
+   *
+   * Allocates a new extent at given existing_paddr that must be absolute and
+   * reads disk to fill the extent.
+   * The common usage is that remove the LogicalCachedExtent (laddr~length at paddr)
+   * and map extent to multiple new extents.
+   * placement_hint and generation should follow the original extent.
+   */
+  using map_existing_extent_iertr =
+    alloc_extent_iertr::extend_ertr<Device::read_ertr>;
+  template <typename T>
+  using map_existing_extent_ret =
+    map_existing_extent_iertr::future<TCachedExtentRef<T>>;
+  template <typename T>
+  map_existing_extent_ret<T> map_existing_extent(
+    Transaction &t,
+    laddr_t laddr_hint,
+    paddr_t existing_paddr,
+    extent_len_t length,
+    placement_hint_t placement_hint = placement_hint_t::HOT,
+    reclaim_gen_t gen = DIRTY_GENERATION) {
+    LOG_PREFIX(TransactionManager::map_existing_extent);
+    if (!existing_paddr.is_absolute() ||
+	!t.is_retired(laddr_hint, length, existing_paddr)) {
+      SUBERRORT(seastore_tm,
+		"existing paddr is not absolute {} or laddr {} not retired",
+		t, existing_paddr, laddr_hint);
+      ceph_abort();
+    }
+
+    auto bp = ceph::bufferptr(buffer::create_page_aligned(length));
+    bp.zero();
+
+    // ExtentPlacementManager::alloc_new_extent will make a new
+    // (relative/temp) paddr, so make extent directly
+    auto ext = CachedExtent::make_cached_extent_ref<T>(std::move(bp));
+
+    ext->init(CachedExtent::extent_state_t::EXIST_CLEAN,
+	      existing_paddr,
+	      placement_hint,
+	      gen);
+
+    t.add_fresh_extent(ext);
+
+    return lba_manager->alloc_extent(
+      t,
+      laddr_hint,
+      length,
+      existing_paddr
+    ).si_then([ext=std::move(ext), laddr_hint, &t, this, FNAME](auto &&ref) {
+      SUBDEBUGT(seastore_tm, "map existing extent: {}, laddr_hint: {} pin: {}",
+		t, *ext, laddr_hint, *ref);
+      ceph_assert(laddr_hint == ref->get_key());
+      ext->set_pin(std::move(ref));
+      return epm->read(
```

Comment:
Is it necessary to actually read it out?

Reply:
@cyx1231st In principal, we could do a lazy read.  I argued, however, for doing this for now in order to simplify the PR.  We can introduce lazy reads later.

Reply:
It's not only about lazy reads. The original data may be already present in the Cache, so no need to read from the disk in that case.

I also agree that it can be in a future PR. The solution might be to pass in the original to-split pin to do splits, that also need refactors in the ObjectDataHandler as it always do removal and insertions in separate rounds.

@zhscn what do you think?

Reply:
> The original data may be already present in the Cache

The details of lazy reads are still incomplete in my mind, but I think lazy reads should handle this case.

> The solution might be to pass in the original to-split pin to do splits, that also need refactors in the ObjectDataHandler as it always do removal and insertions in separate rounds.

It seems reasonable and similar to "write_context_t" that do removals and insertions only once:

> With the "write context" approach, we attach each transaction/ObjectDataHandler with a write_context_t, and cache all steps' results in it, so the workflow changes to the following:
>    1. load all extents that are touched by all transaction ops
>    2. do steps on top of the write_context_t, some of which mean involve extents splitting
>    3. do all retires/allocations

There maybe another way to do lazy read like I have mentioned above is maintaining a structure to record all split operation and the mappings between "split extents" and "(mutated) original extents".

But I'm not convinced it finally works and the implementation looks ugly, since there are several corner cases make things complex.

Reply:
For lazy read, I think the solution may not need to be that complicated. 

After implementing passing in the original to-split pin. Probably to introduce an `ExistCleanPlaceholder` if we found that the data of the to-split pin is not present in Cache (i.e. to-split will be retired as a retired-placeholder) or is also an `ExistCleanPlaceholder`, the `ExistCleanPlaceholder` will be ignored by Cache if we still haven't read it out when commit.

`ExistCleanPlaceholder` will be replaced by a normal EXIST_CLEAN extent if user needs to read it out, so it should never leak to users as well.

Probably we don't need the `ExistCleanPlaceholder` as well, just to allow an EXIST_CLEAN extent to have an empty `ceph::bufferptr` if it is not needed yet.

What do you think?

Reply:
This solution is better and simpler than mine, I'll try to implement it later.

Reply:
> hey @zhscn, are you still working on the lazy read? I am now trying to implement this~

No problem, the lazy read branch last updated a long time ago; I'm not working on it.